core-java/src/main/java/eu/neverblink/jelly/core/utils/IoUtils.java

```diff
@@ -58,4 +59,45 @@ public static void writeFrameAsDelimited(byte[] nonDelimitedFrame, OutputStream
         codedOutput.flush();
         output.write(nonDelimitedFrame);
     }
+
+    /**
+     * Functional interface for processing frames from an input stream.
+     * @param <TFrame> the type of the frame
+     */
+    @FunctionalInterface
+    public interface FrameProcessor<TFrame> {
+        TFrame apply(InputStream inputStream) throws IOException;
+    }
+
+    /**
+     * Reads a stream of frames from an input stream and processes each frame using the provided frame processor.
+     * @param inputStream the input stream to read from
+     * @param frameProcessor the function to process each frame
+     * @param frameConsumer the consumer to handle each processed frame
+     * @param <TFrame> the type of the frame
+     */
+    public static <TFrame> void readStream(
+        InputStream inputStream,
+        FrameProcessor<TFrame> frameProcessor,
+        Consumer<TFrame> frameConsumer
+    ) {
+        while (true) {
+            try {
+                if (inputStream.available() == 0) {
+                    // No more data available, break the loop
+                    break;
+                }
+
+                final var maybeFrame = frameProcessor.apply(inputStream);
+                if (maybeFrame == null) {
+                    // No more frames available, break the loop
+                    break;
+                }
+
+                frameConsumer.accept(maybeFrame);
+            } catch (IOException e) {
+                throw new IllegalStateException("Error reading stream", e);
```

Comment:
We are catching a checked exception here, and it's checked for a reason... so either we rethrow it as a checked exception, or we don't catch it at all. Unless I'm missing something here.